apps/escrow/src/hooks/program/useUploadProgram.ts

```diff
@@ -0,0 +1,195 @@
+  const { docs, method } = api.getExtrinsicFailedError(event);
+  const formattedDocs = docs.filter(Boolean).join('. ');
+
+  return `${method}: ${formattedDocs}`;
+};
+
+const waitForProgramInit = (api: GearApi, programId: string) => {
+  let messageId: Hex;
+  let unsubPromise: UnsubscribePromise;
+
+  const unsubscribe = async () => (await unsubPromise)();
+
+  return new Promise<string>((resolve) => {
+    unsubPromise = api.query.system.events((events) => {
+      events.forEach(({ event }) => {
+        switch (event.method) {
+          case 'MessageEnqueued': {
+            const meEvent = event as MessageEnqueued;
+
+            if (meEvent.data.destination.eq(programId) && meEvent.data.entry.isInit) {
+              messageId = meEvent.data.id.toHex();
+            }
+
+            break;
+          }
+
+          case 'MessagesDispatched': {
+            const mdEvent = event as MessagesDispatched;
+
+            mdEvent.data.statuses.forEach(
+              (status, id) => id.eq(messageId) && status.isFailed && resolve(ProgramStatus.Failed),
+            );
+
+            break;
+          }
+
+          case 'ProgramChanged': {
+            const pcEvent = event as ProgramChanged;
+
+            if (pcEvent.data.id.eq(programId) && pcEvent.data.change.isActive) {
+              resolve(ProgramStatus.Success);
+            }
+
+            break;
+          }
+
+          default:
+            break;
+        }
+      });
+    });
+  }).finally(unsubscribe);
+};
+
+function useUploadProgram(onSuccess: (programId: Hex) => void) {
+  const { api } = useApi();
+  const { account } = useAccount();
+  const alert = useAlert();
+
+  const { uintArray, buffer } = useEscrowOpt();
+  const { metadata } = useEscrowMeta();
+
+  const uploadProgram = async (ftProgramId: Hex) => {
+    if (!account || !buffer || !uintArray || !metadata) return;
+
+    try {
+      const { decodedAddress } = account;
+      const initPayload = { ftProgramId };
+      const value = 0;
+
+      const gas = await api.program.calculateGas.init(decodedAddress, buffer, initPayload, value, false, metadata);
+      const gasLimit = gas.min_limit;
+
+      const alertId = alert.loading('SignIn', { title: 'gear.submitProgram' });
+
+      const { signer } = await web3FromSource(account.meta.source);
+
+      const program = { code: uintArray, gasLimit: gasLimit.toString(), value: value.toString(), initPayload };
+      const { programId } = await api.program.submit(program, metadata, undefined);
+
+      const initialization = waitForProgramInit(api, programId);
+
+      const signStatus = await api.program
+        .signAndSend(account.address, { signer }, ({ status, events }) => {
+          if (status.isReady) {
+            alert.update(alertId, TransactionStatus.Ready);
+
+            return;
+          }
+
+          if (status.isInBlock) {
+            alert.update(alertId, TransactionStatus.InBlock);
+
+            return;
+          }
```

Comment:
You can display status alert using `status.type`. In this case you can avoid if statements

Reply:
But we still need to `return` and alert updates is different in case of `isFinalized` and `isInvalid`, so i don't think that it helps somehow, or am i missing something?

Reply:
okay, as you wish)